### ISSUE TITLE
No callback option for CheckBoxTreeview

### ISSUE THREAD

**Issue by Zolo-mario** [RedFantom/ttkwidgets#22]
so it seems that it is more like a decoration

**Comment by RedFantom** [comment]
If you would like to bind a callback to the `CheckBoxTreeview`, I would like to refer you to the `Treeview` documentation found [here](http://www.tkdocs.com/tutorial/tree.html).

Given how your opening post is only a statement, and you are describing something that is done by design, because there *are* proper ways to make callbacks work with a `Treeview`, I am going to close this issue.

If you would rather like to help improve the widgets in this repository, then please open a *new* issue, describing your exact feature request, and I will take a look at it.

Please also note that this package has not yet reached version number `1.0.0` and is thus not of production quality as of yet.

**Comment by Zolo-mario** [comment]
Sorry for my unthoughtful issue. I will learn more before asking then.

**Comment by WinEunuuchs2Unix** [comment]
Rather than a "callback" I implemented a "call forward" to invoke checkbox processing and immediately validate user results after.

First this line that overrides CheckboxTreeview:

        self.cd_tree.bind('<Button-1>', self.button_1_click)

Then my function which "forward calls" checkbox processing:

    def button_1_click(self, event):
        ''' Call CheckboxTreeview to manage "checked" and "unchecked" tags.
            When it finishes validate what was just checked or unchecked and
            reverse if validation fails.
        '''

        """Check or uncheck box when clicked."""
        # Call CheckboxTreeview function first to let it flag item(s).
        self.cd_tree._box_click(event)

        # Mimic CheckboxTreeview's self._box_click() code
        x, y, widget = event.x, event.y, event.widget
        elem = widget.identify("element", x, y)
        if not "image" in elem: return              # Image was not clicked

        # get item id (iid) matching checkbox that was just clicked
        item = self.cd_tree.identify_row(y)

        # Execute our psuedo-callback functions
        self.reverse_checkbox = False
        if self.cd_tree.tag_has("unchecked", item):
            self.process_unchecked(event, item)
            if self.reverse_checkbox:
                # Call CheckboxTreeview functions to flip checkbox
                self.cd_tree._check_ancestor(item)
                self.cd_tree._check_descendant(item)
        elif self.cd_tree.tag_has("checked", item):
            self.process_checked(event, item)
            if self.reverse_checkbox:
                # Call CheckboxTreeview functions to flip checkbox
                self.cd_tree._uncheck_ancestor(item)
                self.cd_tree._uncheck_descendant(item)
        else:
            # No need to test tristate, item must be checked or unchecked.
            print("button_1_click() ERROR: No 'checked' or 'unchecked' tag.", \
                  self.cd_tree.item(item, 'tags'))
            return

Hope this helps someone in the future.

**Comment by RedFantom** [comment]
@WinEunuuchs2Unix Perhaps I am missing something, but I do not quite understand the general applicability of this code. It cannot be executed by understanding what `reverse_checkbox` means and there appears to be code that is never executed, though of course `process_unchecked(event, item)` may set `self.reverse_chekbox` to `True`, but that is opaque at this point.

I would like to point out that any additional bindings on the `<Button-1>` event can easily be added by considering the standard Tkinter `bind` function, taken from `tkinter/__init__.py`:
```python
class Misc:
    [...]
    def bind(self, sequence=None, func=None, add=None):
        """Bind to this widget at event SEQUENCE a call to function FUNC.
        [...]
        An additional boolean parameter ADD specifies whether FUNC will
        be called additionally to the other bound function or whether
        it will replace the previous function.
        [...]
        """
```
So there is really no need for there to be anything like a 'call forward'. You can just add additional callbacks by doing
```python
cd_tree.bind("<Button-1>", self.my_additional_callback, True)
```

**Comment by WinEunuuchs2Unix** [comment]
@RedFantom Thank you for reviewing the code. I've just changed it for pre-validation of checkbox operations. There is still post-validation of checkbox states however if pee-validation fails the call to `self.cd_tree._box_click(event)` is never made.

I'm still very busy writing new code but constantly go back and revise old code as I learn new "tricks". Your explanation of button hooks requires further study before I can implement it.

BTW thank you for the enhancement you post a few days ago on another question I had: [a490457](https://github.com/TkinterEP/ttkwidgets/commit/a490457c91ca6beed19082a3cb5a57d4b0ea959c). A day after posting the question I came up with a different solution. It doesn't use file `.png` images but rather images in memory.
